164796885-task(authentication): implement jwt authentication

Raymond-Osy opened the pull request:
#### Description
- This PR implements users authentication with JSON web token and also hashes users password

#### Type of change
- New feature (non-breaking change which add functionality)

#### How Has This Been Tested?
- Integration

#### Checklist:
- My code follows the style guideline of this project

#### PT
[#164796885](https://www.pivotaltracker.com/story/show/164796885)

ssanusi committed:
164796882-task(testing) Implemented testing framework
- implement mocha and chai-http
- removed dotenv package

[finishes #164796882]

chikeozulumba committed:
165067148-fix(model): Resolve Model associations

- resolve model association
- fix hound-ci flags

[Fixes #165067148]

Raymond-Osy committed:
164796885-task(authentication): implement jwt authentication
- add bcrypt for hashing user password
- implement the logic for hashing in authenticator.js file
- add json web token for users authentication
- add logic for generating, authenticating and authorising a user
[Delivers 164796885]

Raymond-Osy committed:
164796885-task(authentication): implement jwt authentication
- add bcrypt for hashing user password
- implement the logic for hashing in authenticator.js file
- add json web token for users authentication
- add logic for generating, authenticating and authorising a user
[Delivers 164796885]

Temmyogunbo commented on src/middlewares/authenticator.js:
```diff
@@ -0,0 +1,111 @@
+   * @returns {object} - JSON response object
+   */
+  static isAuthenticated(req, res, next) {
+    const token = req.headers['x-access-token'];
```
Can we try to check the authorization field for our token too?

Raymond-Osy replied:
Noted and implemented, thanks

Temmyogunbo commented on src/middlewares/authenticator.js:
```diff
@@ -0,0 +1,111 @@
+  static isAuthenticated(req, res, next) {
+    const token = req.headers['x-access-token'];
+    if (!token) {
+      res.status(403).json({
```
Do you think this should be 403?

Raymond-Osy replied:
Adjusted, thanks.

Temmyogunbo commented on src/middlewares/authenticator.js:
```diff
@@ -0,0 +1,111 @@
+    } else {
+      jwt.verify(token, process.env.JWTKEY, (err, decoded) => {
+        if (err) {
+          return res.status(403).json({
```
I think 401 is more appropriate because the user ain't authenticated yet.

Raymond-Osy replied:
Thanks for this. Already fixed

Temmyogunbo commented on src/middlewares/authenticator.js:
```diff
@@ -0,0 +1,111 @@
+   * @param {object} next - The next function
+   * @returns {object} - JSON response object
+   */
+  static isAuthenticated(req, res, next) {
```
Can we employ a better name for our method?

Raymond-Osy replied:
Adjusted as recommended, thanks

chikeozulumba committed:
165067148-fix(model): model file structure

- modified .sequelizerc
- moved model files

[Fixes #165067148]

ssanusi committed:
164796882-task(testing) refactor test
- refactor test

[finishes #164796882]

Raymond-Osy committed:
164796885-task(authentication): implement feedback
- add the correct response status
- add utility method for the response format
- move password hashing to the utility file
[Delivers 164796885]

Raymond-Osy committed:
164796885-task(authentication): implement feedback
- add the correct response status
- add utility method for the response format
- move password hashing to the utility file
[Delivers 164796885]

Temmyogunbo commented on src/utils/utils.js:
```diff
@@ -0,0 +1,26 @@
+    return bcrypt.compareSync(password, hashPassword);
+  }
+};
+
```
Let us define these methods in the prototype object of user model.

Temmyogunbo commented on src/middlewares/authenticator.js:
```diff
@@ -0,0 +1,130 @@
+    }
+  }
+}
+
```
You must write tests for all the methods @Raymond-Osy

chikeozulumba committed:
164796891-feature: users can create articles

- add validation function
- create article API
- refactor database config
- add unit testing
- disable sql query logging
- fix query error
- add unit testing
- add integration testing
- update unit testing
- improve article validation

[Delivers #164796891]

chikeozulumba committed:
164796891-feature: users can create articles

- add validation function
- create article API
- refactor database config
- add unit testing
- disable sql query logging
- fix query error
- add unit testing
- add integration testing
- update unit testing
- improve article validation

[Delivers #164796891]

chikeozulumba committed:
164796891-feature: users can create articles

- fix merge conflicts

[Delivers #164796891]

Raymond-Osy committed:
164796885-story(authentication): implement feedback
- add authentication for different roles
- write test case
- add hashing of password in user model
- add compare password method to the utils file
[Delivers 164796885]

Raymond-Osy committed:
164796885-story(authentication): implement feedback
- add authentication for different roles
- write test case
- add hashing of password in user model
- add compare password method to the utils file
[Delivers 164796885]

chikeozulumba committed:
164796891-feature: users can create articles

- refactor tests

[Delivers #164796891]

chikeozulumba committed:
164796891-feature: users can create articles

- refactor article controller

[Delivers #164796891]

chikeozulumba committed:
164796891-feature: users can create articles

- handle errors

[Delivers #164796891]

chikeozulumba committed:
164796891-feature: users can create articles

 - refactor integration tests

[Delivers #164796891]

micah-akpan committed:
#16515309-chore: refactor article / article_rating models

Remove rating field in article model and add value field to article_rating model

[Finishes #16515309]

Temmyogunbo committed:
Merge pull request #36 from andela/story/164796891-user-create-article

164796891 Users Should Be Able To Create Articles

Temmyogunbo commented on src/db_api/models/user.js:
```diff
@@ -65,6 +67,10 @@ export default (sequelize, DataTypes) => {
+
+  User.beforeCreate((user) => {
+    user.prototype.password = bcrypt.hashSync(user.prototype.password, bcrypt.genSaltSync(10));
+  });
```
Please test this properly.

micah-akpan committed:
Add babel-eslint

Temmyogunbo commented on src/middlewares/authenticator.js:
```diff
@@ -0,0 +1,118 @@
+   * @memberOf Authenticator
+   */
+  static generateToken(id) {
+    return jwt.sign({ id }, process.env.JWTKEY, { expiresIn: '24h' });
```
Sync with @Simplemart17 on this.

Temmyogunbo commented on tests/authenticator.test.js:
```diff
@@ -0,0 +1,25 @@
+  });
+  it('should return 401, Unauthorized error', (done) => {
+    agent.get('/api/v1/users/articles')
+      .set('authorization', 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9')
```
We should have the format 'Bearer token"

micah-akpan committed:
Refactor models

[Finishes #165153099]

Temmyogunbo committed:
Merge pull request #40 from andela/chore/165153099-article-model-refactor

#16515309: Refactor Article and Article Rating models

Raymond-Osy committed:
feat(Auth feedback): write more test cases
- implement feedback on jwt authentication
- write more unit test
[Delivers 164796885]

Raymond-Osy committed:
Fix Build issue

Raymond-Osy committed:
Fix build error

Temmyogunbo commented on src/config/databaseConfig.js:
```diff
@@ -3,20 +3,32 @@ module.exports = {
+      ssl: false
+    },
+    username: process.env.DB_USER,
+    logging: false,
```
why this?

Temmyogunbo commented on src/utils/utils.js:
```diff
@@ -0,0 +1,15 @@
+  comparePassword(password, hashPassword) {
+    return bcrypt.compareSync(password, hashPassword);
+  }
+};
```
Remove this.
